crypto/bio/bio_lib.c

```diff
@@ -848,35 +857,44 @@ int BIO_connect_retry(BIO *bio, int timeout)
         return 0;
     }
 
+    if (polling_milliseconds < 0)
+        polling_milliseconds = 100;
     if (!blocking)
         BIO_set_nbio(bio, 1);
 
- retry: /* it does not help here to set SSL_MODE_AUTO_RETRY */
-    rv = BIO_do_handshake(bio); /* This indirectly calls ERR_clear_error(); */
+ retry: /* it does not always help to set SSL_MODE_AUTO_RETRY */
+    rv = BIO_do_connect(bio); /* This may indirectly call ERR_clear_error(); */
 
     if (rv <= 0) {
-        if (get_last_sys_error() == ETIMEDOUT) {
-            /*
-             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
-             */
-            ERR_clear_error();
-            (void)BIO_reset(bio);
-            /*
-             * unless using BIO_reset(), blocking next connect() may crash and
-             * non-blocking next BIO_do_handshake() will fail
-             */
-            goto retry;
-        } else if (BIO_should_retry(bio)) {
-            /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
-            rv = bio_wait(bio, max_time, 100 /* milliseconds */);
+        int err = ERR_peek_last_error();
+        int reason = ERR_GET_REASON(err);
+        int do_retry = BIO_should_retry(bio);
+
+        if (ERR_GET_LIB(err) == ERR_LIB_BIO) {
+            if (reason == BIO_R_NBIO_CONNECT_TIMEOUT
+                    || reason == BIO_R_CONNECT_TIMEOUT) {
+                /* premature timeout has occurred, typically after 30 seconds */
+                (void)BIO_reset(bio); /* needed to avoid failure on retry */
+                ERR_clear_error();
+                do_retry = 1;
+            }
+            else if (reason == ERR_R_SYS_LIB && get_last_sys_error()
+                     == EAGAIN /* resource temporarily unavailable */) {
```

Comment:
I personally, would stop this loop when the address resolution fails.
Technically it is problemati since when we come here, lots of code has run and could clobber errno,
would it be possible to get the original errno from there:
```C
       switch ((gai_ret = getaddrinfo(host, service, &hints, res))) {
# ifdef EAI_SYSTEM
        case EAI_SYSTEM:
            ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
                           "calling getaddrinfo()");
            BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB);
```

Reply:
Meanwhile I see that caling `get_last_socket_error()` is essentially the same as reading `errno`.

In my yesterday's experiments I confirmed that the (previous and current) implementation works fine:
If I use `BIO_do_connect_retry()` in blocking mode or with a sufficiently large timeout value and at first the DNS does not work but after some retries it starts working, the function gets to learn this fact in that way, and the connection succeeds.